### ISSUE TITLE
SimGrid time resolution

### ISSUE THREAD

**Issue by aliomarabdelazim** [mquinson/simgrid#71]
Hello,

I have a question about SIMGrid and I could not find the answer online, What is the time resolution of sigrid, meaning what is the smallest time unit that simgrid simulator can simulate in case of task execution or communication ?
Best regards,
Ali

**Comment by mquinson** [comment]
I see two options that can be related to you question:
```
   surf/precision: Numerical precision used when updating simulation times (in seconds)
       Type: double; Current value: 0.000010
   maxmin/precision: Numerical precision used when computing resource sharing (in ops/sec or bytes/sec)
       Type: double; Current value: 0.000010
```

So, the time resolution is 10-6 seconds, but some tasks may go unnoticed even if they would imply a larger timestamp if the amount of work is too small (that could happen on a very slow resource).

**Comment by mquinson** [comment]
Actually, I think that this question is not really an issue in our code base, and the best would be to post it on stack overflow, that offers a better support for this kind of things. @aliomarabdelazim , would you mind reposting your question there, please?

Thanks in advance, Mt

**Comment by aliomarabdelazim** [comment]
Thank you for your reply and consideration,
However the issue is that I have been trying to figure out the the time resolution by trying to simulate the execution of tasks of 10 FLOPs on 1 GFLOP/s hosts, and I can not see difference in the printed execution time when running with default precision or using --cfg=surf/precision:1-e9 to increase the precision, and it is not clear to me is this a printing limitation (I am using XBT_INFO to print the simulation time) or really the execution time is zero and the option did not increase the precision,
I am using Simgrid 3.13 and using SIMDAG interface,

thank you for your help
Best regards,
Ali

Ali Omar Abdelazim Mohammed
Departement Mathematik & Informatik
High Performance Computing Group
Universität Basel
Spiegelgasse 1
CH-4051 Basel Switzerland
email: ali.mohammed@unibas.ch<mailto:ali.mohammed@unibas.ch>
Phone: +41 (0)61 267 05 41

On 06 Apr 2016, at 21:27, Martin Quinson <notifications@github.com<mailto:notifications@github.com>> wrote:


Actually, I think that this question is not really an issue in our code base, and the best would be to post it on stack overflow, that offers a better support for this kind of things. @aliomarabdelazim<https://github.com/aliomarabdelazim> , would you mind reposting your question there, please?

Thanks in advance, Mt

**Comment by mquinson** [comment]
Hello,
It seems to us that you mainly have a logging issue, not a precision one. 

The first thing to try is to change the logging format using the information of http://simgrid.gforge.inria.fr/simgrid/3.13/doc/group__XBT__log.html#log_use_conf_fmt
In short, you should try adding ```--log=root.fmt:[%12.9r]%e(%i:%P@%h)%e%m%n``` on the command line. The difference with the default setting is that usually we have ```%12.6r```, requesting for the date to be displayed with a precision of 10-6 second.

Another thing if this does not suffice to try would be to send 1000000 such very small tasks to see whether they sum up to something or not.

HTH

**Comment by aliomarabdelazim** [comment]
thank you very much, this was very helpful. the problem was actually a logging issue as you suspected. However it seems like I can not experimentally find the exact time resolution of SIMGRID, I tried with hosts of “10000000000000000Ef” speed and task of 1 flop and still get execution time “0.00000000000000000000000000000000010000”  which is impressive. can you suggest me an experiment to try to find the time resolution ?
Best regards,
Ali



On 07 Apr 2016, at 22:19, Martin Quinson <notifications@github.com<mailto:notifications@github.com>> wrote:


Hello,
It seems to us that you mainly have a logging issue, not a precision one.

The first thing to try is to change the logging format using the information of http://simgrid.gforge.inria.fr/simgrid/3.13/doc/group__XBT__log.html#log_use_conf_fmt
In short, you should try adding --log=root.fmt:[%12.9r]%e(%i:%P@%h)%e%m%n on the command line. The difference with the default setting is that usually we have %12.6r, requesting for the date to be displayed with a precision of 10-6 second.

Another thing if this does not suffice to try would be to send 1000000 such very small tasks to see whether they sum up to something or not.

HTH

**Comment by frs69wq** [comment]
You're trying to execute 1 floating point operation on a machine running 
at 10^15 Exaflops (that is 10^33 flop/s)! The time you obtain has 33 
digits before the 1, so this looks correct. Completely unrealistic with 
regard to the current processing power of the most powerful 
supercomputers in the world, but logical in terms of simulation. You ask 
for "How much time does it take to compute 1 operation on such a huge 
machine", you get that time. The only limit is your imagination:-)

Le 14/04/2016 09:41, aliomarabdelazim a écrit :
> thank you very much, this was very helpful. the problem was actually a 
> logging issue as you suspected. However it seems like I can not 
> experimentally find the exact time resolution of SIMGRID, I tried with 
> hosts of “10000000000000000Ef” speed and task of 1 flop and still get 
> execution time “0.00000000000000000000000000000000010000” which is 
> impressive. can you suggest me an experiment to try to find the time 
> resolution ?
> Best regards,
> Ali
>
>
>
> On 07 Apr 2016, at 22:19, Martin Quinson 
> <notifications@github.com<mailto:notifications@github.com>> wrote:
>
>
> Hello,
> It seems to us that you mainly have a logging issue, not a precision one.
>
> The first thing to try is to change the logging format using the 
> information of 
> http://simgrid.gforge.inria.fr/simgrid/3.13/doc/group__XBT__log.html#log_use_conf_fmt
> In short, you should try adding 
> --log=root.fmt:[%12.9r]%e(%i:%P@%h)%e%m%n on the command line. The 
> difference with the default setting is that usually we have %12.6r, 
> requesting for the date to be displayed with a precision of 10-6 second.
>
> Another thing if this does not suffice to try would be to send 1000000 
> such very small tasks to see whether they sum up to something or not.
>
> HTH
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on 
> GitHub<https://github.com/mquinson/simgrid/issues/71#issuecomment-207073186>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/mquinson/simgrid/issues/71#issuecomment-209809096>
>

-- 
One should, for example, be able to see that things are hopeless
and yet be determined to make them otherwise.
                                           -- F. Scott Fitzgerald

**Comment by mquinson** [comment]
So, can we close this issue maybe?

**Comment by aliomarabdelazim** [comment]
I am still trying to understand how sigrid steps the simulation and how it advances time, so what I find currently strange is,
I tried to schedule 16 tasks on 4 hosts each host is 1 Exa flop of power, and 8 tasks is of 1 flop and the other 8 is of 10000000000000000 flops, the simulation time I got is 0.02000000000000000388578058618804789148, and when I removed the 8 big tasks and only scheduled 8 tasks of 1 flops I got execution time of 0.00000000000000000200000000000000014308, and a third time I only scheduled the big tasks I got execution times of
0.02000000000000000041633363423443370266, and the numbers do not sum to each other, is there an explanation for this


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "simgrid/simdag.h"
#include "xbt/log.h"
#include "xbt/ex.h"



XBT_LOG_NEW_DEFAULT_CATEGORY(test, "Logging specific to this SimDag example");

typedef struct _HostAttribute *HostAttribute;
struct _HostAttribute {
  /* Earliest time at which a host is ready to execute a task */
  double available_at;
  SD_task_t last_scheduled_task;
};

static void sg_host_allocate_attribute(sg_host_t host)
{
  void *data;
  data = calloc(1, sizeof(struct _HostAttribute));
  sg_host_user_set(host, data);
}

static void sg_host_free_attribute(sg_host_t host)
{
  free(sg_host_user(host));
  sg_host_user_set(host, NULL);
}


static void sg_host_set_available_at(sg_host_t host, double time)
{
  HostAttribute attr = (HostAttribute) sg_host_user(host);
  attr->available_at = time;
  sg_host_user_set(host, attr);
}

static SD_task_t sg_host_get_last_scheduled_task( sg_host_t host){
  HostAttribute attr = (HostAttribute) sg_host_user(host);
  return attr->last_scheduled_task;
}

static void sg_host_set_last_scheduled_task(sg_host_t host, SD_task_t task){
  HostAttribute attr = (HostAttribute) sg_host_user(host);
  attr->last_scheduled_task=task;
  sg_host_user_set(host, attr);
}




int main(int argc, char **argv)
{
  unsigned int cursor;
  SD_task_t task, selected_task = NULL, last_scheduled_task;
  sg_host_t selected_host = NULL;
  int total_nhosts = 0;
  int numTasks = 16;
  const sg_host_t *hosts = NULL;
  SD_task_t taskArray[numTasks];
  int numHosts=0;
  sg_host_t *executedHosts;
  xbt_dynar_t changed_tasks;
  char taskName[10];
  /* initialization of SD */
  SD_init(&argc, argv);

  /* Check our arguments */
  xbt_assert(argc > 1, "Usage: %s platform_file \n"
             "\tExample: %s platform_hosts.xml ", argv[0], argv[0]);


  /* creation of the environment */
  SD_create_environment(argv[1]);

  /*  Allocating the host attribute */
  total_nhosts = sg_host_count();
  hosts = sg_host_list();

//creating tasks
  XBT_INFO("Creating Tasks");
  for (int i = 0; i < 8; i++) {
    sprintf(taskName,"Task %d", i);
    taskArray[i]=SD_task_create_comp_seq(taskName, NULL, 1);
    //SD_task_watch(taskArray[i], SD_DONE);
  }
  for (int i = 8; i < 16; i++) {
    sprintf(taskName,"Task %d", i);
    taskArray[i]=SD_task_create_comp_seq(taskName, NULL, 10000000000000000);
    //SD_task_watch(taskArray[i], SD_DONE);
  }
  XBT_INFO("Finished Creating tasks");
  XBT_INFO("Creating data elements for hosts");
  for (cursor = 0; cursor < total_nhosts; cursor++)
    sg_host_allocate_attribute(hosts[cursor]);
  XBT_INFO("finished creating data elements for hosts");

  XBT_INFO("Scheduling Tasks on hosts--distributing tasks among hosts");
  //Schedule tasks on hosts ...3 tasks for every host
  for (size_t i = 0; i < total_nhosts; i++) {
    for (size_t j = 0; j < numTasks/total_nhosts; j++)
    {
      selected_host=hosts[i];
      selected_task=taskArray[i + total_nhosts*j];
      SD_task_schedulel(selected_task, 1, selected_host);
      XBT_INFO("Task ''%s' Scheduled on Host ''%s' ", SD_task_get_name (selected_task),SD_workstation_get_name (selected_host));
      /*
       * SimDag allows tasks to be executed concurrently when they can by default.
       * Yet schedulers take decisions assuming that tasks wait for resource availability to start.
       * The solution (well crude hack is to keep track of the last task scheduled on a host and add a special type of
       * dependency if needed to force the sequential execution meant by the scheduler.
       * If the last scheduled task is already done, has failed or is a predecessor of the current task, no need for a
       * new dependency
      */

      last_scheduled_task = sg_host_get_last_scheduled_task(selected_host);
      if (last_scheduled_task && (SD_task_get_state(last_scheduled_task) != SD_DONE) &&
          (SD_task_get_state(last_scheduled_task) != SD_FAILED) &&
          !SD_task_dependency_exists(sg_host_get_last_scheduled_task(selected_host), selected_task))
        SD_task_dependency_add("resource", NULL, last_scheduled_task, selected_task);

      sg_host_set_last_scheduled_task(selected_host, selected_task);
      sg_host_set_available_at(selected_host,SD_task_get_finish_time(selected_task));
    }
  }

  changed_tasks = SD_simulate(-1.0);
  xbt_dynar_foreach(changed_tasks, cursor, task) {
    XBT_INFO("Task '%s' start time: %f, finish time: %f", SD_task_get_name(task), SD_task_get_start_time(task),
             SD_task_get_finish_time(task));
    numHosts = SD_task_get_workstation_count(task);
    executedHosts = SD_task_get_workstation_list(task);
    for (int i = 0; i < numHosts; i++) {
      XBT_INFO("Task '%s' executed on %s ",SD_task_get_name(task),SD_workstation_get_name(executedHosts[i]) );
    }

  }


  XBT_INFO("Simulation Time: %7.24f", SD_get_clock());



xbt_dynar_foreach(changed_tasks, cursor, task) {
  SD_task_destroy(task);
}

  for (cursor = 0; cursor < total_nhosts; cursor++)
    sg_host_free_attribute(hosts[cursor]);

  /* exit */
  SD_exit();
  return 0;
}

Best regards,
Ali



On 16 Apr 2016, at 12:31, Martin Quinson <notifications@github.com<mailto:notifications@github.com>> wrote:


So, can we close this issue maybe?

**Comment by frs69wq** [comment]
Hi, 

The processing power if a host is shared evenly between the tasks running on it. When 16 tasks are executed on the same host at the same time they get 1/16th of the power each. It won't be for a long time as 8 of them just need to compute one flop, but it will slow down a bit the big ones. 

I didn't do the math (I'm on my phone waiting for a tramway) but the time when you run everything should be 1exaflop/16 + (the compute amount -1)/8.

This is different of 1exaflop/8 +the compute amount/8.

Have a look to our 101 tutorials on the simgrid website, one of them at least explains how resource sharing is computed.

F.

Envoyé de mon iPhone

**Comment by aliomarabdelazim** [comment]
Thank you very much for your fast reply, I forgot to mention that these tasks are executed on four hosts (each of power 1 exa flops )and I create dependencies between tasks allocated on the same host to avoid resource sharing, you can see the code attached below, also here is what my platform file looks like

Best regards,
Ali


<?xml version='1.0'?>
<!DOCTYPE platform SYSTEM "http://simgrid.gforge.inria.fr/simgrid/simgrid.dtd">
<platform version="4">
<AS id="main" routing="Full">
        <link sharing_policy="SHARED" bandwidth="8Gbps" latency="500ns" id="link1" state="ON"></link>

    <host core="1" pstate="0.0" id="host 1" speed="1Ef" availability="1.0" state="ON"></host>
<host core="1" pstate="0.0" id="host 2" speed="1Ef" availability="1.0" state="ON"></host>
<host core="1" pstate="0.0" id="host 3" speed="1Ef" availability="1.0" state="ON"></host>
<host core="1" pstate="0.0" id="host 4" speed="1Ef" availability="1.0" state="ON"></host>


<route src="host 1" dst="host 2"><link_ctn id="link1" direction="NONE"></link_ctn></route>
<route src="host 1" dst="host 3"><link_ctn id="link1" direction="NONE"></link_ctn></route>
<route src="host 1" dst="host 4"><link_ctn id="link1" direction="NONE"></link_ctn></route>

</AS>
</platform>


On 18 Apr 2016, at 20:30, Frederic Suter <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi,

The processing power if a host is shared evenly between the tasks running on it. When 16 tasks are executed on the same host at the same time they get 1/16th of the power each. It won't be for a long time as 8 of them just need to compute one flop, but it will slow down a bit the big ones.

I didn't do the math (I'm on my phone waiting for a tramway) but the time when you run everything should be 1exaflop/16 + (the compute amount -1)/8.

This is different of 1exaflop/8 +the compute amount/8.

Have a look to our 101 tutorials on the simgrid website, one of them at least explains how resource sharing is computed.

F.

Envoyé de mon iPhone

**Comment by mquinson** [comment]
Did you read the following tutorial? It may help understanding the simgrid internals.
http://simgrid.gforge.inria.fr/tutorials/surf-101.pdf
